libs/ml/examples/word2vec_distributed_learning/translator.hpp

```diff
@@ -0,0 +1,130 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/byte_array/const_byte_array.hpp"
+#include "core/digest.hpp"
+#include "math/tensor.hpp"
+#include "ml/dataloaders/word2vec_loaders/vocab.hpp"
+
+#include <utility>
+#include <vector>
+
+namespace fetch {
+namespace ml {
+namespace distributed_learning {
+
+struct Translator
+{
+  using SizeType = fetch::math::SizeType;
+
+  template <typename TensorType>
+  std::pair<TensorType, TensorType> Translate(TensorType                        gradient_update,
+                                              const byte_array::ConstByteArray &vocab_hash)
+  {
+    using DataType = typename TensorType::Type;
+
+    TensorType mask = TensorType({MyVocabSize()});
+
+    if (vocab_hash == MyVocabHash())  // no need to translate
+    {
+      mask.Fill(DataType{1});
+      return std::make_pair(gradient_update, mask);
+    }
+
+    assert(VocabKnown(vocab_hash));
+
+    std::vector<std::string> other_vocab = known_vocabs[vocab_hash];
+
+    // figure out which way around the matrix is
+    bool       first_dim_embedding;
+    SizeType   embedding_size;
+    TensorType transl_gu;
```

Comment:
its not obvious to me what a transl_gu is.
might as well rename the variable to something written in full here